networkmanager/networkmanager.go

```diff
@@ -472,6 +464,19 @@ func (manager *NetworkManager) SetTrafficPeriod(period int) error {
  * Private
  **********************************************************************************************************************/
 
+func (manager *NetworkManager) deleteInstanceNetworkFromCache(instanceID, networkID string) error {
+	manager.Lock()
```

Comment:
There is a race condition here: in `AddInstanceToNetwork` you have add instance to network and before locking on mutex `RemoveInstanceFromNetwork` is called. It will delete the network. As result `AddInstanceToNetwork` will fail.

Reply:
The same race condition is still possible: for example new instance is added after unlock. The proper solution is to lock the whole function. It will not have to much impact on starting or stopping 100 instances as it will be blocked only once.